File: examples/canonical-terminal/ruby-thor-tty/lib/taskflow/core/job.rb
# frozen_string_literal: true

require "time"

module Taskflow
  module Core
    Job = Struct.new(:id, :name, :status, :started_at, :duration_s, :tags, :output_lines, keyword_init: true) do
      def self.from_h(payload)
        new(
          id: payload.fetch("id"),
          name: payload.fetch("name"),
          status: payload.fetch("status"),
          started_at: parse_time(payload["started_at"]),
          duration_s: payload["duration_s"],
          tags: Array(payload["tags"]),
          output_lines: Array(payload["output_lines"])
        )
      end

      def to_h
        {
          id: id,
          name: name,
          status: status,
          started_at: self.class.format_time(started_at),
          duration_s: duration_s,
          tags: tags,
          output_lines: output_lines
        }
      end

      def self.parse_time(value)
        value.nil? ? nil : Time.iso8601(value).utc
      end

      def self.format_time(value)
        value&.utc&.iso8601
      end
    end

    Event = Struct.new(:event_id, :job_id, :event_type, :timestamp, :message, keyword_init: true) do
      def self.from_h(payload)
        new(
          event_id: payload.fetch("event_id"),
          job_id: payload.fetch("job_id"),
          event_type: payload.fetch("event_type"),
          timestamp: Time.iso8601(payload.fetch("timestamp")).utc,
          message: payload.fetch("message")
        )
      end

      def to_h
        {
          event_id: event_id,
          job_id: job_id,
          event_type: event_type,
          timestamp: timestamp.utc.iso8601,
          message: message
        }
      end
    end
  end
end

File: examples/canonical-terminal/elixir-optimus-ratatouille/lib/taskflow/core/job.ex
defmodule Taskflow.Core.Job do
  @enforce_keys [:id, :name, :status, :tags, :output_lines]
  defstruct [:id, :name, :status, :started_at, :duration_s, :tags, :output_lines]

  def from_map(payload) do
    %__MODULE__{
      id: Map.fetch!(payload, "id"),
      name: Map.fetch!(payload, "name"),
      status: Map.fetch!(payload, "status"),
      started_at: parse_time(Map.get(payload, "started_at")),
      duration_s: Map.get(payload, "duration_s"),
      tags: Map.get(payload, "tags", []),
      output_lines: Map.get(payload, "output_lines", [])
    }
  end

  def to_map(%__MODULE__{} = job) do
    %{
      id: job.id,
      name: job.name,
      status: job.status,
      started_at: format_time(job.started_at),
      duration_s: job.duration_s,
      tags: job.tags,
      output_lines: job.output_lines
    }
  end

  defp parse_time(nil), do: nil
  defp parse_time(value), do: DateTime.from_iso8601(value) |> elem(1)
  defp format_time(nil), do: nil
  defp format_time(value), do: DateTime.to_iso8601(value)
end

defmodule Taskflow.Core.Event do
  @enforce_keys [:event_id, :job_id, :event_type, :timestamp, :message]
  defstruct [:event_id, :job_id, :event_type, :timestamp, :message]

  def from_map(payload) do
    {:ok, timestamp, _offset} = DateTime.from_iso8601(Map.fetch!(payload, "timestamp"))

    %__MODULE__{
      event_id: Map.fetch!(payload, "event_id"),
      job_id: Map.fetch!(payload, "job_id"),
      event_type: Map.fetch!(payload, "event_type"),
      timestamp: timestamp,
      message: Map.fetch!(payload, "message")
    }
  end
end
